banyand/measure/measure_topn.go

```diff
@@ -0,0 +1,568 @@
+			filters, buildErr := manager.buildFilters(conditions)
+			if buildErr != nil {
+				return buildErr
+			}
+			streamingFlow = streamingFlow.Filter(filters)
+		}
+
+		mapper, innerErr := manager.buildMapper(topNSchema.GetFieldName(), topNSchema.GetGroupByTagNames()...)
+		if innerErr != nil {
+			return innerErr
+		}
+		streamingFlow = streamingFlow.Map(mapper)
+
+		processor := &topNStreamingProcessor{
+			l:                manager.l,
+			shardNum:         manager.m.shardNum,
+			interval:         interval,
+			topNSchema:       topNSchema,
+			databaseSupplier: manager.m.databaseSupplier,
+			src:              srcCh,
+			in:               make(chan interface{}),
+			streamingFlow:    streamingFlow,
+		}
+
+		processor.errCh = streamingFlow.Window(streaming.NewSlidingTimeWindows(interval, slideSize)).
+			TopN(int(topNSchema.GetCountersNumber()),
+				streaming.WithSortKeyExtractor(func(elem interface{}) int64 {
+					return elem.(api.Data)[1].(int64)
+				}),
+				streaming.OrderBy(topNSchema.GetFieldValueSort()),
+			).To(processor).OpenAsync()
+
+		manager.processorMap[topNSchema.GetSourceMeasure()] = processor
+	}
+
+	return nil
+}
+
+func (manager *topNProcessorManager) buildFilters(criteria []*modelv1.Criteria) (measureProcessorFilter, error) {
+	var filters []conditionFilter
+	for _, group := range criteria {
+		for _, cond := range group.GetConditions() {
+			fIdx, tIdx, spec := pbv1.FindTagByName(manager.m.GetSchema().GetTagFamilies(), cond.GetName())
+			if spec == nil {
+				return nil, errors.New("fail to parse tag by name")
+			}
+			filters = append(filters, manager.buildFilterForTag(fIdx, tIdx, cond))
+		}
+	}
+
+	if len(filters) == 0 {
+		return func(value *measurev1.DataPointValue) bool {
+			return true
+		}, nil
+	}
+
+	return func(dataPoint *measurev1.DataPointValue) bool {
+		tfs := dataPoint.GetTagFamilies()
+		for _, f := range filters {
+			if !f.predicate(tfs) {
+				return false
+			}
+		}
+		return true
+	}, nil
+}
+
+func (manager *topNProcessorManager) buildFilterForTag(familyOffset, tagOffset int, cond *modelv1.Condition) conditionFilter {
+	switch v := cond.GetValue().GetValue().(type) {
+	case *modelv1.TagValue_Int:
+		return &int64TagFilter{
+			TagLocator: partition.TagLocator{
+				FamilyOffset: familyOffset,
+				TagOffset:    tagOffset,
+			},
+			op:  cond.GetOp(),
+			val: v.Int.GetValue(),
+		}
+	case *modelv1.TagValue_Str:
+		return &strTagFilter{
+			TagLocator: partition.TagLocator{
+				FamilyOffset: familyOffset,
+				TagOffset:    tagOffset,
+			},
+			op:  cond.GetOp(),
+			val: v.Str.GetValue(),
+		}
+	case *modelv1.TagValue_Id:
+		return &idTagFilter{
+			TagLocator: partition.TagLocator{
+				FamilyOffset: familyOffset,
+				TagOffset:    tagOffset,
+			},
+			op:  cond.GetOp(),
+			val: v.Id.GetValue(),
+		}
+	default:
+		return nil
+	}
+}
+
+func (manager *topNProcessorManager) buildMapper(fieldName string, groupByNames ...string) (measureProcessorMapper, error) {
```

Comment:
Currently, `groupByNames` is required. Shall we consider a pre-`TopN Aggregation` without groupBy?

This would affect the layout we save for the TopN "virtual" measure.

Reply:
Could you map data into a single group in case of an missing group? If that, you don't have to refactor too many points.

Reply:
> Could you map data into a single group in case of an missing group? If that, you don't have to refactor too many points.

To be clear, you meant mapping all data into a single group if groupByNames is not specified, right?